### PR TITLE
LSTM hidden states

### PR DESCRIPTION
**Pull request by rusane**
- LSTMs now use the last hidden state from the batch at the start of the next batch.
- Added reset method for LSTM hidden states, which is called in `optimize()` of `optim.py`.
- Add batch normalization between LSTM and FC layer (not completely sure if this is the correct way).
- Added dependency parameter to the model (default=independent).

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -10,19 +10,31 @@
...
+    def __init__(self, num_const, num_var, num_hidden, num_out, num_groups, num_batch, independent):
         super(SubModel, self).__init__()
-        self.lstm = SplitLSTM(num_const, num_var, num_hidden, num_groups, independent=True) # TODO make independent default
-        self.fc = SplitLinear(0, num_hidden, num_out, num_groups) # TODO add independent parameter for consistency
+        self.num_hidden = num_hidden
+        self.num_groups = num_groups
+        self.num_batch = num_batch
+
+        self.lstm = SplitLSTM(num_const, num_var, num_hidden, num_groups, independent)
+        self.bn = nn.BatchNorm1d(num_groups)
```
Comment (nnistelrooij) [inline model.py:16]: The batch size is always equal to 1. Batch Normalization is then not effective as the running statistics are based on one sample. Batch Normalization is only worthwhile with a sufficiently large batch size. There are modified versions of BN that do work with RNNs: https://arxiv.org/pdf/1603.09025.pdf with code: https://github.com/jihunchoi/recurrent-batch-normalization-pytorch/blob/master/bnlstm.py
Maybe take a look at those (I just searched for the paper and the code for 5 minutes), for a normalization layer that works better with LSTMs specifically.

File: model.py [hunk 2]
```diff
@@ -10,19 +10,31 @@
...
+        self.fc = SplitLinear(0, num_hidden, num_out, num_groups, independent)
+
+        self.hidden = (torch.zeros(1, num_batch, num_hidden*num_groups),
+                       torch.zeros(1, num_batch, num_hidden*num_groups))
+                    
+    def reset_hidden(self):
+        self.hidden = (torch.zeros(1, self.num_batch, self.num_hidden*self.num_groups),
```
Comment (nnistelrooij) [inline model.py]: It does not work on the GPU, probably because self.hidden is not initialized on the GPU.

Comment (nnistelrooij) [reply model.py]: A more maintainable way of fixing this would be to save the hidden state in `SplitLSTM`. Then when you call `reset_hidden()`, you set `self.hidden` to `None`. If you pass a `hx` that is equal to `None` to `self.lstm`, the layer will initialize the hidden state on the correct device automatically. This makes more sense, because the hidden state is more a part of the LSTM than of the entire sub-model.

File: model.py (hunk 1 above)
Comment (rusane) [reply model.py:16]: Thanks, I'll look into it some more, I think I actually found that code too when working on it.

File: model.py (hunk 2 above)
Comment (rusane) [reply model.py]: Will do. The reason why I did it this way is because I regarded `SplitLSTM` as a replacement of `nn.LSTM`. As such, `hidden` would be specified in the module where you assign the LSTM layer normally.

Comment (nnistelrooij) [reply model.py]: It is a shame that the hidden state needs to be detached from the computational graph every iteration. This means that backpropagation through time (BPTT) will only work for a maximum of `seq_len` time steps. But alas, hopefully by having `seq_len > 1`, the training efficiency will still be increased compared to not having a hidden state.

Comment (rusane) [reply model.py]: I moved the hidden state to `SplitLSTM` and it works on GPU now (note: you still have to change device in forward of `mlps.py` to `cuda`). GPU is much slower compared to CPU though from what I've seen.

File: model.py (hunk 1 above)
Comment (rusane) [reply model.py:16]: It seems it is not really straightforward to use BN with LSTMs. We would have to reparameterize the LSTM like has been done for the `LSTMCell` in the code you linked. I honestly have no clue how to do that and I'm not sure if it is worth the time for us to figure it out.

Should I take out the `BatchNorm1d` for now at least as it is not really useful as you said with a batch size of 1?

Comment (nnistelrooij) [reply model.py:16]: OK. Then that will be a task for next week after we've uploaded a submission on Kaggle on the trained LSTM models.
